stan/math/rev/core/chainable_object.hpp

```diff
@@ -0,0 +1,67 @@
+#ifndef STAN_MATH_REV_CORE_CHAINABLE_OBJECT_HPP
+#define STAN_MATH_REV_CORE_CHAINABLE_OBJECT_HPP
+
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/core/chainable_alloc.hpp>
+#include <stan/math/rev/core/typedefs.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * `chainable_object` hold another object is useful for connecting
+ * the lifetime of a specific object to the chainable stack
+ *
+ * `chainable_object` objects should only be allocated with `new`.
+ * `chainable_object` objects allocated on the stack will result
+ * in a double free (`obj_` will get destructed once when the
+ * chainable_object leaves scope and once when the chainable
+ * stack memory is recovered).
+ *
+ * @tparam T type of object to hold
+ */
+template <typename T>
+class chainable_object : public chainable_alloc {
+ private:
+  T obj_;
```

Comment:
I'd put `std::decay_t<T>` here since I think we always want the object to hold a non-reference type

Reply:
Or just `plain_type_t<T>`

Reply:
Talked to @t4c1 about this earlier: https://github.com/stan-dev/math/pull/2230#discussion_r535910828 and we went with decay (also we did this at the `make_chainable_ptr` level).

But now that you bring it up again, I think I want to reverse the decision and do `plain_type`.

Maybe `chainable_object` can have a `decay` in it, but when I do `make_chainable_ptr(x)`, I want to make it so that `x` is available on the reverse-mode pass, whatever `x` is, without worrying about the template type (that's why I made the function). Much easier to avoid mistakes here if I just `plain_type` it.

I think I could make a use where I need to save expressions that reference arena memory, but then I could just copy the expression by value and not need this at all.

Reply:
Fine, I guess. I can't really think of an example where we would want to store a non-plain type.

Reply:
Hahaha, my laziness wins. `plain_type_t` is so nice cause then I don't have to think.